### PR TITLE
Use of unified rule engine(ure) for queries

### PR DESCRIPTION
**Pull request by noskill**
1) Split method to generate query in questionToOpencog converter to support both: ure and pattern matcher queries.
2) Use ure in gui demo

### CODE REVIEW

File: experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py [hunk 1]
```diff
@@ -200,7 +242,9 @@ def addBoundingBoxesIntoAtomspace(self, features):
             imageFeatures = FloatValue(boundingBoxFeatures)
             boundingBoxInstance = ConceptNode(
                 'BoundingBox-' + str(boundingBoxNumber))
-            InheritanceLink(boundingBoxInstance, ConceptNode('BoundingBox'))
+            inh = InheritanceLink(boundingBoxInstance, ConceptNode('BoundingBox'))
+            tv = TruthValue(0.999, 0.999)
```
Comment (vsbogd) [inline experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py]: Why 0.999? :-)

File: experiments/opencog/question2atomese/src/main/java/org/opencog/vqa/relex/WhatOtherDetObjSubjToSchemeQueryConverter.java [hunk 2]
```diff
@@ -28,22 +38,31 @@ public String getSchemeQuery(RelexFormula relexFormula) {
...
-                "    (EvaluationLink (GroundedPredicateNode \"py:runNeuralNetwork\") (ListLink (VariableNode \"$B\") (VariableNode \"$X\")) )\n" +
-                "  )\n" +
+                "  )\n" + andLink + 
                 "  (ListLink (Variable \"$B\") (Variable \"$X\") (ConceptNode \"%2$s\"))\n" +
                 ")\n"
                 , visitor.attribute, visitor.object);
     }
     
+    @Override
+    public String getSchemeQueryURE(RelexFormula relexFormula) {
```
Comment (vsbogd) [inline experiments/opencog/question2atomese/src/main/java/org/opencog/vqa/relex/WhatOtherDetObjSubjToSchemeQueryConverter.java:52]: I would add unit test for this method

File: experiments/opencog/question2atomese/src/main/java/org/opencog/vqa/relex/YesNoPredadjToSchemeQueryConverter.java [hunk 3]
```diff
@@ -27,16 +27,36 @@ public String getSchemeQuery(RelexFormula relexFormula) {
...
                 "    (EvaluationLink (GroundedPredicateNode \"py:runNeuralNetwork\") (ListLink (VariableNode \"$X\") (ConceptNode \"%1$s\")) )\n" +
                 "    (EvaluationLink (GroundedPredicateNode \"py:runNeuralNetwork\") (ListLink (VariableNode \"$X\") (ConceptNode \"%2$s\")) )\n" +
-                "  )\n" + 
+                "  )\n" +
+                "  (VariableNode \"$X\")" +
                 ")\n", visitor.object, visitor.state);
     }
 
+    @Override
+    public String getSchemeQueryURE(RelexFormula relexFormula) {
```
Comment (vsbogd) [inline experiments/opencog/question2atomese/src/main/java/org/opencog/vqa/relex/YesNoPredadjToSchemeQueryConverter.java:46]: I would add unit test for this method

File: experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py (hunk 1 above)
Comment (noskill) [reply experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py]: I am quite sure that code above produces InheritanceLink, but not 100% sure - there could be bug or something. I believe that probabilities of zero and one are applicable only to logical statements.

Comment (noskill) [reply experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py]: I'll try 1.0 0.999. Default stv(1, 0) gives stv(1, 0) for queries to URE with fuzzy logic formulas with AndLinks

File: experiments/opencog/question2atomese/src/main/java/org/opencog/vqa/relex/YesNoPredadjToSchemeQueryConverter.java (hunk 3 above)
Comment (noskill) [reply experiments/opencog/question2atomese/src/main/java/org/opencog/vqa/relex/YesNoPredadjToSchemeQueryConverter.java:46]: done

File: experiments/opencog/question2atomese/src/main/java/org/opencog/vqa/relex/WhatOtherDetObjSubjToSchemeQueryConverter.java (hunk 2 above)
Comment (noskill) [reply experiments/opencog/question2atomese/src/main/java/org/opencog/vqa/relex/WhatOtherDetObjSubjToSchemeQueryConverter.java:52]: Done

File: experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py [hunk 4]
```diff
@@ -345,35 +420,37 @@ def answerOtherQuestion(self, queryInScheme):
...
-                    continue
-                results.append(OtherDetSubjObjConjunction(strength, confidence, predicate_name))
+                bounding_box_id = extract_bb_id(out)
+                if a_extract_predicate:
+                    predicate_name = extract_predicate(out)
+                    if not predicate_name:
+                        continue
+                else:
+                    predicate_name = None
+                results.append(ConjunctionResult(strength, confidence, predicate_name, bounding_box_id, resultData))
```
Comment (vsbogd) [inline experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py]: I would propose keep only resultData in ConjunctionResult and calculate strength, confidence, predicate_name and bounding_box_id from resultData on the fly.
```python
class ConjunctionResult(OtherDetSubjObj):
    def __init__(resultData):
        self.resultData = resultData

    # getter or property
    def strength(self):
        return resultData.tv.mean
```

File: experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py [hunk 5]
```diff
@@ -345,35 +420,37 @@ def answerOtherQuestion(self, queryInScheme):
...
-            for atom in atoms:
-                if atom.type == opencog.atomspace.types.InheritanceLink:
-                    if atom.out[1].name != 'BoundingBox':
-                        predicate_name = atom.out[0].name
-                        return predicate_name
+        results = self.sort_results(resultsData, a_extract_predicate=True)
+        if not results:
+            return None, None, None
+        maxResult = results[0]
+        answer = maxResult.get_answer()
```
Comment (vsbogd) [inline experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py]: minor, It seems you don't need local variables you can just write:

    return maxResult.get_answer(), maxResult.get_bounding_box_id(), maxResult.get_expression()

File: experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py [hunk 6]
```diff
@@ -305,7 +366,7 @@ def answerQuestion(self, record, use_pm=True):
                 self.logger.error('Question was not parsed')
                 return
             self.logger.debug('Scheme query: %s', queryInScheme)
-            answer = self.answerQuery(record.questionType, queryInScheme)
+            answer, bb_id, expr = self.answerQuery(record.questionType, queryInScheme)
```
Comment (vsbogd) [inline experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py]: minor, but I would propose replace bb_id and expr by ```_``` if you don't need them it reads better

File: experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py [hunk 7]
```diff
@@ -191,11 +223,13 @@ def get_answer(self):
         return self.attribute.name
 
 
-class OtherDetSubjObjConjunction(OtherDetSubjObj):
-    def __init__(self, strength, confidence, predicate_name):
+class ConjunctionResult(OtherDetSubjObj):
```
Comment (vsbogd) [inline experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py:226]: See https://github.com/singnet/semantic-vision/pull/12/commits/7108ea6a64f1841931a8d1282785a0214dc3079c#r222984367 you can extract all values from resultData in the class,

**Review by vsbogd — APPROVED** [review]

File: experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py (hunk 7 above)
Comment (noskill) [reply experiments/opencog/pattern_matcher_vqa/pattern_matcher_vqa.py:226]: Done

**Pull request merged**